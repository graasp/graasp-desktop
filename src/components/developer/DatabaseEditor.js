import React, { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';
import ReactJson from 'react-json-view';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { withTranslation } from 'react-i18next';
import { withStyles } from '@material-ui/core';
import { getDatabase, setDatabase } from '../../actions';
import Loader from '../common/Loader';
import Styles from '../../Styles';
import SampleDatabase from '../../data/sample.json';

export class DatabaseEditor extends Component {
  static propTypes = {
    t: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      button: PropTypes.string.isRequired,
    }).isRequired,
    dispatchGetDatabase: PropTypes.func.isRequired,
    dispatchSetDatabase: PropTypes.func.isRequired,
    database: PropTypes.shape({
      user: PropTypes.object,
      spaces: PropTypes.array,
    }),
  };

  static defaultProps = {
    database: {},
  };

  componentDidMount() {
    const { dispatchGetDatabase } = this.props;
    dispatchGetDatabase();
  }

  handleEdit = ({ updated_src: updatedSrc }) => {
    const { dispatchSetDatabase } = this.props;
    dispatchSetDatabase(updatedSrc);
  };

  handleUseSampleDatabase = () => {
    const { dispatchSetDatabase } = this.props;
    dispatchSetDatabase(SampleDatabase);
  };

  render() {
    const { database, t, classes } = this.props;

    if (!database || _.isEmpty(database)) {
      return <Loader />;
    }

    return (
      <div>
        <Typography variant="h6">{t('Manually Edit the Database')}</Typography>
        <ReactJson
          collapsed
          src={database}
          onEdit={this.handleEdit}
          onAdd={this.handleEdit}
          onDelete={this.handleEdit}
        />
        <br />
        <Button
          variant="contained"
          className={classes.button}
          onClick={this.handleUseSampleDatabase}
          color="primary"
        >
          {t('Use Sample Database')}
        </Button>
      </div>
    );
  }
}

const mapStateToProps = ({ Developer }) => ({
  database: Developer.get('database'),
});

const mapDispatchToProps = {
  dispatchGetDatabase: getDatabase,
  dispatchSetDatabase: setDatabase,
};

const StyledComponent = withStyles(Styles, { withTheme: true })(DatabaseEditor);
const TranslatedComponent = withTranslation()(StyledComponent);
const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(TranslatedComponent);

export default ConnectedComponent;
