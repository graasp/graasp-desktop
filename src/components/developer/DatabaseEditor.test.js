import React from 'react';
import { shallow } from 'enzyme';
import ReactJson from 'react-json-view';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { DatabaseEditor } from './DatabaseEditor';
import Loader from '../common/Loader';

const sampleDatabase = {
  user: { name: 'name' },
  spaces: [{}],
};

const createDatabaseEditorProps = database => {
  return {
    t: text => text,
    classes: {
      button: '',
    },
    dispatchGetDatabase: jest.fn(),
    dispatchSetDatabase: jest.fn(),
    database,
  };
};

describe('<DatabaseEditor />', () => {
  describe('<DatabaseEditor /> renders', () => {
    let props;
    let wrapper;

    beforeAll(() => {
      props = createDatabaseEditorProps(sampleDatabase);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<DatabaseEditor {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <Typography /> component with text "Manually Edit the Database"', () => {
      const typography = wrapper.find(Typography);
      expect(typography).toHaveLength(1);
      expect(typography.contains('Manually Edit the Database')).toBeTruthy();
    });

    it('renders one <Button /> component with text "Use Sample Database"', () => {
      const button = wrapper.find(Button);
      expect(button).toHaveLength(1);
      expect(button.contains('Use Sample Database')).toBeTruthy();
    });

    it('renders one <ReactJson /> component with sample database', () => {
      const reactJson = wrapper.find(ReactJson);
      expect(reactJson).toHaveLength(1);
      expect(reactJson.prop('src')).toEqual(props.database);
    });
  });

  describe.each([undefined, {}])(
    '<DatabaseEditor /> on undefined or empty database',
    database => {
      let wrapper;

      it('renders <Loader />', () => {
        const props = createDatabaseEditorProps(database);
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<DatabaseEditor {...props} />);
        expect(wrapper.find(Loader)).toHaveLength(1);
      });
    }
  );
});
