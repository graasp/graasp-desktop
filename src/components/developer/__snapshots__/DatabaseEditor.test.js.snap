// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatabaseEditor /> <DatabaseEditor /> renders renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Typography))
    variant="h6"
  >
    Manually Edit the Database
  </WithStyles(ForwardRef(Typography))>
  <t
    collapseStringsAfterLength={false}
    collapsed={true}
    defaultValue={null}
    displayDataTypes={true}
    displayObjectSize={true}
    enableClipboard={true}
    groupArraysAfterLength={100}
    iconStyle="triangle"
    indentWidth={4}
    name="root"
    onAdd={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onSelect={false}
    shouldCollapse={false}
    sortKeys={false}
    src={
      Object {
        "spaces": Array [
          Object {},
        ],
        "user": Object {
          "name": "name",
        },
      }
    }
    style={Object {}}
    theme="rjv-default"
    validationMessage="Validation Error"
  />
  <br />
  <WithStyles(ForwardRef(Button))
    className=""
    color="primary"
    onClick={[Function]}
    variant="contained"
  >
    Use Sample Database
  </WithStyles(ForwardRef(Button))>
</div>
`;
